windows/RNPrint/RNPrint.h

```diff
@@ -0,0 +1,402 @@
+  {
+    std::optional<std::string> html = std::nullopt;
+    std::optional<std::string> filePath = std::nullopt;
+    bool isLandscape = false;
+    std::string jobName;
+  };
+
+  REACT_MODULE(RNPrint);
+  struct RNPrint
+  {
+    const std::string Name = "RNPrint";
+
+    // Default name for the print job.
+    inline static const std::string defaultJobName = "Document";
+
+    // Name for the XAML Canvas to add preview folders to.
+    inline static const std::string RNPrintCanvasName = "RNPrintCanvas";
+
+    // These are needed between print PDF callbacks, so we place them in the struct.
+    ReactContext reactContext = nullptr;
+    Canvas printCanvas = nullptr; // holds reference to RNPrintCanvas in visual tree
+    PrintPageDescription printPageDescr;
+    PrintDocument printDoc = nullptr;
+    IPrintDocumentSource printDocSource;
+    int numberOfPages = 0;
+    std::map<int, std::optional<xaml::UIElement>> pageCollection; // references to the xaml pages
+    std::mutex printSync; // prevent race conditions when manipulating pageCollection or checking a print request is being processed
+    bool servingPrintRequest = false;
+
+    // Event token for print registratio.
+    winrt::event_token printToken;
+
+    REACT_INIT(RNPrint_Init)
+      void RNPrint_Init(ReactContext const& context) noexcept
+    {
+      reactContext = context;
+    }
+
+    // Searches for the RNPrintCanvas element in the visual tree.
+    Canvas searchForPrintCanvas(xaml::DependencyObject startNode)
+    {
+      int count = winrt::Windows::UI::Xaml::Media::VisualTreeHelper::GetChildrenCount(startNode);
+      Canvas result = nullptr;
+      for (int i = 0; i < count; i++)
+      {
+        xaml::DependencyObject current = winrt::Windows::UI::Xaml::Media::VisualTreeHelper::GetChild(startNode, i);
+        Canvas temp = current.try_as<Canvas>();
+        if (temp)
+        {
+          std::string tempName = winrt::to_string(temp.Name());
+          if (tempName == RNPrintCanvasName)
+          {
+            return temp;
+          }
+        } else
+        {
+          result = searchForPrintCanvas(current);
+          if (result) break;
+        }
+      }
+      return result;
+    }
+
+    // cleans up local variables and XAML elements.
+    void cleanUp()
+    {
+      reactContext.UIDispatcher().Post([=]()
+        {
+          auto printMan = PrintManager::GetForCurrentView();
+          {
+            std::lock_guard<std::mutex> guard(printSync);
+            servingPrintRequest = false;
+            if (printToken)
+              printMan.PrintTaskRequested(printToken);
+            printToken = {};
+            pageCollection.clear();
+            if (printCanvas)
+              printCanvas.Children().Clear();
+          }
+        }
+      );
+    }
+
+    // Generates a XAML page asynchronously.
+    concurrency::task<std::optional<xaml::UIElement> > GeneratePageAsync(PdfDocument pdfDocument, int pageNumber, PrintPageDescription pageDescription)
```

Comment:
nit: you shouldn't need the space between > > anymore, I think? :)

Reply:
muscle memory from times past :)